metrics/mahalanobis/mahalanobis.py

```diff
@@ -0,0 +1,97 @@
+# Copyright 2021 The HuggingFace Datasets Authors and the current dataset script contributor.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Mahalanobis metric."""
+
+import numpy as np
+
+import datasets
+
+
+_DESCRIPTION = """
+Compute the Mahalanobis Distance
+
+Mahalonobis distance is the distance between a point and a distribution.
+And not between two distinct points. It is effectively a multivariate equivalent of the Euclidean distance.
+It was introduced by Prof. P. C. Mahalanobis in 1936
+and has been used in various statistical applications ever since
+[source: https://www.machinelearningplus.com/statistics/mahalanobis-distance/]
+"""
+
+_CITATION = """\
+@article{de2000mahalanobis,
+  title={The mahalanobis distance},
+  author={De Maesschalck, Roy and Jouan-Rimbaud, Delphine and Massart, D{\'e}sir{\'e} L},
+  journal={Chemometrics and intelligent laboratory systems},
+  volume={50},
+  number={1},
+  pages={1--18},
+  year={2000},
+  publisher={Elsevier}
+}
+"""
+
+_KWARGS_DESCRIPTION = """
+Args:
+    predictions: Predicted labels, as returned by a model.
+    references: Ground truth labels.
+Returns:
+    matthews_correlation: Matthews correlation.
+Examples:
+
+    >>> mahalanobis_metric = datasets.load_metric("mahalanobis")
+    >>> results = mahalanobis_metric.compute(references=[[0, 1], [1, 0]], predictions=[[0, 1]])
```

Comment:
`predictions` must be a single list right ?
```suggestion
    >>> results = mahalanobis_metric.compute(references=[[0, 1], [1, 0]], predictions=[0, 1])
```

Reply:
No, it can be a list of vectors. It will yield the mahalanobis distance for each of those input vectors.

Reply:
It seems that this kind of metric is not supported due to [this exception](https://github.com/huggingface/datasets/blob/207be676bffe9d164740a41a883af6125edef135/src/datasets/metric.py#L469). Do you suggest for me to change the function to only accept a single list? It will be slower to compute prediction by prediction

Reply:
Indeed your `references` parameter is actually the definition of a distribution, not references values for each prediction. In this case, please remove `references` from the features and `_compute`, and add the `reference_distribution` parameter to `_compute`.

The example will have to be changed to
```python
results = mahalanobis_metric.compute(reference_distribution=[[0, 1], [1, 0]], predictions=[[0, 1]])
```

Under the hood it means that `predictions` is the input of the metric, and `reference_distribution` simply becomes a parameter of the metric.

(feel free to rename `predictions` and `reference_distribution` if you want btw, you can use whatever variable names make more sense)

Reply:
Makes sense 👍 Thought it was a hard argument naming constraint that I had to follow.

Changed `predictions` -> `X` and `references` -> `reference_distribution` in https://github.com/huggingface/datasets/pull/3794/commits/2037075ce881901a79883880938a792362735fca